### ISSUE TITLE
Extend unnecessary_lazy_evaluations to more combinators

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust-clippy#7814]
The current program currently doesn't trigger any warnings, but should trigger `unnecessary_lazy_evaluations` due to the use of `map_or_else` instead of `map_or`:

```rust
const FOO: Option<u16> = Some(1);

fn main() {
    let val = FOO.map_or_else(|| 0, |v| v + 1);
    println!("val: {}", val);
}
```

@rustbot label +C-enhancement
